### ISSUE TITLE
Bar chart bars overlapping when data is a little spread out

### ISSUE THREAD

**Issue by iron-man** [c3js/c3#669]
Bar chart bars overlap each other when the data isn't as nicely spread out, but sporadic.  Here's a fiddle of the issue. http://jsfiddle.net/Ljennqjw/1/ 
notice how the bars on the right all overlap.

**Comment by raulgomis** [comment]
same for me! is there a workaround to avoid that? @masayuki0812 any update on this? thanks

**Comment by raulgomis** [comment]
I think it should calculate the min witdth that allows to fit all the data and the use this width for all the bars.

**Comment by iron-man** [comment]
What I am doing right now for this is to just set the size of all bars to be really small, so they only overlap if the screen is super small, even though it makes it look strange for larger display, they don't overlap.  Still hoping this gets add/fixed soon.

**Comment by raulgomis** [comment]
Another workaround is to calculate the days that are missing in the timeseries and to set it to zero. It should work if you don't mind when is zero or the data is missing. That's the problem on this workaround.

**Comment by raulgomis** [comment]
I have checked and works if you create the missing days and set the data to null, they appear in the axes but the tooltip does not display as there are no data, so it is a great solution. It would be great, apart from fixing this issue, to have an option in the timeseries to display the missing days, i.e. to calculate them and show them in the axes with empty data. What do you think? @masayuki0812

**Comment by raulgomis** [comment]
The plunkr here: http://plnkr.co/edit/aYLWQXtz3mKzu4iMhwK3

**Comment by iron-man** [comment]
Yeah, but if the data is more sparse, such as in this modified plunkr http://plnkr.co/edit/vOyBZCda0pTUMXQQmRr2?p=info then you see the same thing, are you saying that I will have to have fill in that entire time-series with null values?  Seems like it should just handle null values on the back-end gracefully.

**Comment by raulgomis** [comment]
Yes, it is the workaround...to fill the missing headers and fill the data with null values.

**Comment by malefice** [comment]
I am a newbie to D3 and C3, so please feel free to shoot down or to agree with my ideas, but I rather share them so C3 can become even much better! :D

Anyway, chances are the data to be loaded in a timeseries (including the missing/null columns) will have a regular interval. In the plunkr samples, the common interval type would still be a day, even though most of the dates were missing. If worse comes to worst, the interval type would be a minute or even a second, but I highly doubt a lot of cases will resort to that kind of granularity.

On that note, I think adding an "interval type" and "interval size" for timeseries to compute the widths of the columns would be a great feature! If interval type is minute and the size is 43, that means an interval is 43 mins. The number of intervals would then be the difference between the max and min datetimes divided by 43. From there, one should be able to compute the correct width. Did I make sense? XD

**Comment by masayuki0812** [comment]
Hi, Sorry for late response, but I think `bar.width` can be used although it's not fit automatically and we need to set it properly. If automatic computation is needed, it seems better to mark enhancement.

**Comment by ramdog** [comment]
Hi, I stumbled across this recently as I faced a similar issue with bars, timeseries, and uneven date intervals between data points.  @masayuki0812's suggestion to use `bar.width` certainly works. I have another possible suggestion:

I believe the problem is that default calculation of the bar width is the following in pseudo code:
```
bar width  = width of plot area / number of items in data * scaling factor
```
... where the scaling factor is the the default [bar.width.ratio](http://c3js.org/reference.html#bar-width-ratio) of __0.6__ which makes things look pretty.  The issue is that the denominator above should actually be the number of tick marks (or date intervals) we would expect along the date range, not the number of data points included in the sporadically spaced data. (And this is challenging to make as an enhancement because how is `c3.js` going to know what the interval should be, i.e. days, weeks, months, though I guess it could be configured.)

However, because the `bar.width.ratio` can be overridden, and because we as the developer _know_ the basis for the interval, we can solve for what it should be to achieve the same default behavior /  spacing.  The calculation boils down to `bar.width.ratio = x / y * 0.6` where `x` is the number of items in the unevenly spaced data points, `y` is how many items should have been considered (for the denominator from above) and of course `0.6` is the default `bar.width.ratio`.

So, I updated @iron-man's original fiddle to demonstrate this (I also took the liberty of rotating the labels).
http://jsfiddle.net/Ljennqjw/26/

Now, to get the tick marks at every interval, not just the data points...

**Comment by zhu-7** [comment]
Hello @ramdog, I am using your workaround for plotting timeseries bar chart with unevenly spaced dates. But I do have one question, is there a way to compute the ticks instead of hard coded? The magic number 28, how is it computed? 
I would assume to get the timespan between the first and last date from the data, but since x Axis contains buffer before first date and after last date, would be better to use the min/max date from x Axis, right? 
Is there a cleaner way to make it work? Thanks a lot for your help.

**Comment by ramdog** [comment]
Hi @zhu-7.  To answer your questions:
1. The magic number 28 in that example is the items that _should have been considered_ in dataset - in that case it was # of - the months between start and end.  It could (should) be computed yes.
2. Not sure if there is a cleaner solution. At the time of writing, I don't recall seeing a better technique.  The challenging part is that the library (C3) doesn't know if you want to plot months / years / days etc.  So computing that denominator would be hard.  Maybe there is a better way to do this now.
